# DUBStepR
DUBStepR (Determining the Underlying Basis using Step-wise Regression) is a feature selection algorithm for cell type identification in single-cell RNA-sequencing data.


## Release Notes
Version 1.0 released on 10 December 2019.

## Setting Up DUBStepR

### Installing DUBStepR

DUBStepR requires your R version to be >= 3.5.0. Once you've ensured that, you can install DUBStepR from GitHub using the following commands:

```R
install.packages("devtools")
devtools::install_github("bbbranjan/DUBStepR")
```

### Loading DUBStepR into your environment

After installation, load DUBStepR using the following command:

```R
library(DUBStepR)
```

## Using DUBStepR

For users new to single-cell RNA sequencing data analysis, we recommend using DUBStepR with the [Seurat](https://satijalab.org/seurat/) (Satija, R. et al. Nature Biotechnol. 33.5(2015):495.) package to analyze single-cell RNA seq data. Below is a tutorial using the Seurat package.

### Install and load Seurat

Seurat can be installed and loaded into your R environment using the following commands:

```R
install.packages("Seurat")
library(Seurat)
library(dplyr)
```

### Prepare Seurat object

Here, we use a publicly available PBMC dataset generated by 10X Genomics. Here's a [link](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_v3) to the dataset. We use the **Feature / cell matrix HDF5 (filtered)** file.

Locate the file in your working directory and load the data in to your Seurat object in the following manner:

```R
seuratObj <- CreateSeuratObject(counts = Read10X_h5("pbmc_10k_v3_filtered_feature_bc_matrix.h5"), assay = "RNA", project = "10k_PBMC")
seuratObj[["percent.mt"]] <- PercentageFeatureSet(seuratObj, pattern = "^MT-")
seuratObj <- SubsetData(object = seuratObj, subset.name = "percent.mt", low.threshold = -Inf, high.threshold = 10) 
seuratObj <- SubsetData(object = seuratObj, subset.name = "nFeature_RNA", low.threshold = 200, high.threshold = Inf)
```

For DUBStepR, we recommend log-normalizing your 10X data. That can be performed in Seurat using the following command:

```R
seuratObj <- NormalizeData(object = seuratObj, normalization.method = "LogNormalize")
```


### Run DUBStepR to identify feature genes

DUBStepR can be inserted into the Seurat workflow at this stage, and we recommend that be done in the following manner:

```R
seuratObj@assays$RNA@var.features <- DUBStepR(input.data = seuratObj@assays$RNA@data, min.cells = 100, k = 10, num.pcs = 15)
```
```
[1] "Expression Filtering - Done"
[1] "Mitochondrial, Ribosomal and Pseudo Genes Filtering - Done"
[1] "kNN Smoothing - Done"
---> Splitting data matrix: 9 splits of 11769x1204 size
---> Splitting data matrix: 1 split of 11769x1211 size
[1] "Running Stepwise Regression"
  |=============================================================================================================================| 100%[1] "Determining optimal feature set"
  |=============================================================================================================================| 100%
```


### Visualize and cluster cells

We first visualize the cells on a 2D UMAP projection

```R
seuratObj <- ScaleData(seuratObj, features = rownames(seuratObj))
seuratObj <- RunPCA(seuratObj, features = VariableFeatures(object = seuratObj))
ElbowPlot(seuratObj)
seuratObj <- FindNeighbors(seuratObj, reduction = "pca", dims = 1:12)
seuratObj <- FindClusters(seuratObj, resolution = 0.2)
seuratObj <- RunUMAP(seuratObj, dims = 1:12)
DimPlot(seuratObj, reduction = "umap")
cell.types <- c("Monocytes", "CD4+ T cells", "B cells", "CD8+ T cells", "Lymphoc", "NK cells", "FCGR3A+ Monocytes", "Platelets", "DC", "Unknown", "mDC", "Progenitors")
```

![](images/Cluster_Plot.png "Cluster Plot")

We select the first few feature genes selected by DUBStepR to show cell type specific expression
```R
FeaturePlot(seuratObj, features = VariableFeatures(object = seuratObj)[1:12])
```

![](images/DUBStepR_Feature_Plot.png "DUBStepR Feature Plot")

Using known marker genes, we show cell type specific regions of the UMAP
```R
FeaturePlot(seuratObj, features = c("MS4A1", "GNLY", "NKG7", "CD3E", "CD14", "FCER1A", "CST3", "FCGR3A", "LYZ", "PPBP", "CD8A", "IL7R", "CCR7", "S100A4", "MS4A7", "CX3CR1"))
```

![](images/Known_Marker_Feature_Plot.png "Known Markers Feature Plot")


Since DUBStepR predicts cell type specific feature genes, we can look at the overlap with DE genes of these clusters to check its accuracy


```R
all.markers <- FindAllMarkers(seuratObj, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 2)
unique.markers <- unique(all.markers$gene)
100*sum(markers %in% VariableFeatures(object = seuratObj))/length(unique.markers)
```

```
[1] 70.54264
```
